src/H5FDint.c

```diff
@@ -736,211 +996,201 @@ H5FD_read_selection(H5FD_t *file, uint32_t count, H5FD_mem_t type, H5S_t *mem_sp
-                                HGOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, FAIL,
-                                            "memory reallocation failed for buffer list")
-                            vec_bufs = tmp_ptr;
-                        }
-
-                        /* Record that we've doubled the array sizes */
-                        vec_arr_nalloc *= 2;
-                    }
+    if (file->base_addr > 0) {
 
-                    /* Add this segment to vector write list */
-                    addrs[vec_arr_nused]    = offsets[i] + file_off[file_seq_i];
-                    sizes[vec_arr_nused]    = io_len;
-                    vec_bufs[vec_arr_nused] = (void *)((uint8_t *)buf + mem_off[mem_seq_i]);
-                    vec_arr_nused++;
-                }
-                else
-                    /* Issue scalar read call */
-                    if ((file->cls->read)(file, type, dxpl_id, offsets[i] + file_off[file_seq_i], io_len,
-                                          (void *)((uint8_t *)buf + mem_off[mem_seq_i])) < 0)
-                    HGOTO_ERROR(H5E_VFL, H5E_READERROR, FAIL, "driver read request failed")
-
-                /* Update file sequence */
-                if (io_len == file_len[file_seq_i])
-                    file_seq_i++;
-                else {
-                    file_off[file_seq_i] += io_len;
-                    file_len[file_seq_i] -= io_len;
-                }
+        /* apply the base_addr offset to the offsets array.  Must undo before
+         * we return.
+         */
+        for (i = 0; i < count; i++) {
 
-                /* Update memory sequence */
-                if (io_len == mem_len[mem_seq_i])
-                    mem_seq_i++;
-                else {
-                    mem_off[mem_seq_i] += io_len;
-                    mem_len[mem_seq_i] -= io_len;
-                }
+            offsets[i] += file->base_addr;
+        }
+        offsets_cooked = TRUE;
+    }
 
-                /* Refill file sequence list if necessary */
-                if (file_seq_i == H5FD_SEQ_LIST_LEN) {
-                    if (H5S_SELECT_ITER_GET_SEQ_LIST(&file_iter, H5FD_SEQ_LIST_LEN, SIZE_MAX, &file_nseq,
-                                                     &dummy_nelem, file_off, file_len) < 0)
-                        HGOTO_ERROR(H5E_INTERNAL, H5E_UNSUPPORTED, FAIL, "sequence length generation failed")
+    /* If the file is open for SWMR read access, allow access to data past
+     * the end of the allocated space (the 'eoa').  This is done because the
+     * eoa stored in the file's superblock might be out of sync with the
+     * objects being written within the file by the application performing
+     * SWMR write operations.
+     */
+    /* For now at least, only check that the offset is not past the eoa, since
+     * looking into the highest offset in the selection (different from the
+     * bounds) is potentially expensive.
+     */
```

Comment:
Assuming we shift to VFD SWMR, the possibility of reading past EOA in the SWMR case will go away, as VFD SWMR should always provide the reader with the correct EOA.

We will need a good way of testing to see if a selection read request extends beyond EOA.